hilti/toolchain/src/compiler/cfg.cc

```diff
@@ -0,0 +1,893 @@
+                            return contains(*scope_end->scope, *decl);
+                        });
+
+                        if ( is_local_only )
+                            continue;
+                    }
+
+                    auto [_, inserted] = in.insert(f);
+                    changed |= inserted;
+                }
+            }
+
+            // The out set of a node is gen + (in - kill)
+            const auto& gen = _dataflow.at(n).gen;
+            const auto& kill = _dataflow.at(n).kill;
+
+            for ( auto&& [decl, g] : gen ) {
+                auto [_, inserted] = out.insert(g);
+                changed |= inserted;
+            }
+
+            for ( auto&& i : in ) {
+                if ( std::any_of(kill.begin(), kill.end(), [&](auto&& kv) {
+                         auto&& [_k, n] = kv;
+                         return n.count(i);
+                     }) )
+                    continue;
+
+                auto [_, inserted] = out.insert(i);
+                changed |= inserted;
+            }
+        }
+
+        if ( ! changed )
+            break;
+    }
+}
+
+std::vector<Node*> CFG::unreachableStatements() const {
+    // This can only be called after reachability information has been populated.
+    assert(! _dataflow.empty());
+    assert(_dataflow.begin()->second.reachability);
+
+    std::map<Node*, uint64_t> uses;
+
+    // Loop over all nodes.
+    for ( const auto& [n, transfer] : _dataflow ) {
+        // Check whether we want to declare any of the statements used. We currently do this for
+        // - `inout` parameters since their result is can be seen after the function has ended,
+        // - globals since they could be used elsewhere without us being able to see it,
+        // - `self` expression since they live on beyond the current block.
+        if ( n->isA<End>() ) {
+            assert(_dataflow.count(n));
+            // If we saw an operation an `inout` parameter at the end of the flow, mark the parameter as used.
+            // For each incoming statement ...
+            for ( auto&& in : transfer.reachability->in ) {
+                assert(_dataflow.count(in));
+                // If the statement generated an update to the value ...
+                for ( auto&& [n_, _] : _dataflow.at(in).gen ) {
+                    if ( n_->isA<declaration::GlobalVariable>() )
+                        ++uses[in];
+
+                    else if ( auto&& p = n_->tryAs<declaration::Parameter>();
+                              p && p->kind() == parameter::Kind::InOut ) {
+                        ++uses[in];
+                    }
+
+                    else if ( const auto* expr = n_->tryAs<declaration::Expression>() ) {
+                        if ( auto* keyword = expr->expression()->tryAs<expression::Keyword>();
+                             keyword && keyword->kind() == expression::keyword::Kind::Self )
+                            ++uses[in];
```

Comment:
Is this run after the aliases are marked as uses? If so I think a case like this would fail (if aliases were good on locals, regardless it gets an incorrect result now):

```
module Test {

import hilti;

type S = struct {
    uint<64> i;

    method void print() {
        hilti::print(self.i);
    }

    method void f();
} &on-heap;

method void S::f() {
    self.print();
    local strong_ref<S> ref = self;
    ref.i = 1;
    self.print();
    ref.i = 2;
    self.print();
}

global S s = [$i = 0];
s.f();

}
```

should print `0 1 2`, prints `0 0 0`